File: packages/sdk/src/modules.ts
import { DefarmHttpClient } from "./client.js";
import {
  AuthResponse,
  Circuit,
  DisclosureRequest,
  DisclosureResponse,
  Event,
  Item,
  PartnerIntakeResponse,
  ReceiptSummary,
} from "./types.js";

export class AuthApi {
  constructor(private readonly http: DefarmHttpClient) {}

  async login(email: string, password: string): Promise<AuthResponse> {
    return this.http.request<AuthResponse>("POST", "/auth/login", { email, password });
  }

  async refresh(refreshToken: string): Promise<AuthResponse> {
    return this.http.request<AuthResponse>("POST", "/auth/refresh", { refresh_token: refreshToken });
  }

  async whoami(): Promise<AuthResponse["user"]> {
    return this.http.request<AuthResponse["user"]>("GET", "/auth/me");
  }

  async logout(refreshToken?: string): Promise<{ message?: string }> {
    return this.http.request<{ message?: string }>("POST", "/auth/logout", {
      refresh_token: refreshToken,
    });
  }
}

export class WorkspaceApi {
  constructor(private readonly http: DefarmHttpClient) {}

  async list() {
    return this.http.request<{ workspaces: unknown[]; count: number }>("GET", "/auth/workspaces");
  }

  async status() {
    return this.http.request<AuthResponse["user"]>("GET", "/auth/me");
  }
}

export class CircuitsApi {
  constructor(private readonly http: DefarmHttpClient) {}

  async list(): Promise<Circuit[]> {
    const res = await this.http.request<{ circuits?: Circuit[] } | Circuit[]>("GET", "/api/circuits");
    return Array.isArray(res) ? res : res.circuits || [];
  }

  async show(id: string): Promise<Circuit> {
    return this.http.request<Circuit>("GET", `/api/circuits/${id}`);
  }

  async members(id: string) {
    return this.http.request("GET", `/api/circuits/${id}/members`);
  }

  async join(id: string, message?: string) {
    return this.http.request("POST", `/api/circuits/${id}/join-requests`, message ? { message } : {});
  }
}

export class ItemsApi {
  constructor(private readonly http: DefarmHttpClient) {}

  async list(circuitId?: string): Promise<Item[]> {
    const qs = circuitId ? `?circuit_id=${encodeURIComponent(circuitId)}` : "";
    const res = await this.http.request<{ items?: Item[] } | Item[]>("GET", `/api/items${qs}`);
    return Array.isArray(res) ? res : res.items || [];
  }

  async show(id: string) {
    return this.http.request("GET", `/api/items/${id}`);
  }

  async create(payload: Record<string, unknown>) {
    return this.http.request("POST", "/api/items", payload);
  }

  async createViaIngestion(payload: {
    source_circuit_id: string;
    items: Array<Record<string, unknown>>;
    fallback_to_source_circuit?: boolean;
    auto_create_circuit?: boolean;
    mapping?: Record<string, unknown>;
  }): Promise<PartnerIntakeResponse> {
    return this.http.request<PartnerIntakeResponse>("POST", "/v1/partner/ingestions", payload);
  }

  async update(id: string, payload: Record<string, unknown>) {
    return this.http.request("PUT", `/api/items/${id}`, payload);
  }
}

export class EventsApi {
  constructor(private readonly http: DefarmHttpClient) {}

  async list(circuitId?: string): Promise<Event[]> {
    const qs = circuitId ? `?circuit_id=${encodeURIComponent(circuitId)}` : "";
    const res = await this.http.request<{ events?: Event[] } | Event[]>("GET", `/api/events${qs}`);
    return Array.isArray(res) ? res : res.events || [];
  }

  async show(id: string): Promise<Event> {
    return this.http.request<Event>("GET", `/api/events/${id}`);
  }

  async add(payload: Record<string, unknown>) {
    return this.http.request("POST", "/api/events", payload);
  }

  async update(id: string, payload: Record<string, unknown>) {
    return this.http.request("PUT", `/api/events/${id}/status`, payload);
  }
}

export class DisclosuresApi {
  constructor(private readonly http: DefarmHttpClient) {}

  async create(payload: DisclosureRequest): Promise<DisclosureResponse> {
    return this.http.request<DisclosureResponse>("POST", "/api/disclosures", payload);
  }
}

export class ReceiptsApi {
  constructor(private readonly http: DefarmHttpClient) {}

  async list(params?: {
    receipt_type?: string;
    circuit_id?: string;
    item_id?: string;
    limit?: number;
    offset?: number;
  }): Promise<ReceiptSummary[]> {
    const qs = new URLSearchParams();
    if (params?.receipt_type) qs.set("receipt_type", params.receipt_type);
    if (params?.circuit_id) qs.set("circuit_id", params.circuit_id);
    if (params?.item_id) qs.set("item_id", params.item_id);
    if (params?.limit != null) qs.set("limit", String(params.limit));
    if (params?.offset != null) qs.set("offset", String(params.offset));
    const suffix = qs.toString() ? `?${qs.toString()}` : "";
    const res = await this.http.request<{ receipts?: ReceiptSummary[] } | ReceiptSummary[]>("GET", `/api/receipts${suffix}`);
    return Array.isArray(res) ? res : res.receipts || [];
  }

  async show(id: string) {
    return this.http.request("GET", `/api/receipts/${id}`);
  }
}

File: rust/defarm-sdk/src/modules.rs
use serde::Deserialize;
use serde_json::{json, Value};

use crate::client::DefarmClient;
use crate::error::Result;
use crate::types::{
    Circuit, DisclosureRequest, DisclosureResponse, Event, EventInput, Item, ItemIngestionInput,
    PartnerIntakeResponse, ReceiptSummary,
};

#[derive(Deserialize)]
#[serde(untagged)]
enum CollectionResponse<T> {
    Wrapped {
        #[serde(alias = "items", alias = "circuits", alias = "events", alias = "receipts")]
        items: Vec<T>,
    },
    Bare(Vec<T>),
}

impl<T> CollectionResponse<T> {
    fn into_inner(self) -> Vec<T> {
        match self {
            Self::Wrapped { items } => items,
            Self::Bare(items) => items,
        }
    }
}

pub struct CircuitsApi {
    client: DefarmClient,
}

impl CircuitsApi {
    pub(crate) fn new(client: DefarmClient) -> Self {
        Self { client }
    }

    pub async fn list(&self) -> Result<Vec<Circuit>> {
        let resp: CollectionResponse<Circuit> = self.client.get("/api/circuits").await?;
        Ok(resp.into_inner())
    }

    pub async fn show(&self, id: &str) -> Result<Circuit> {
        self.client.get(&format!("/api/circuits/{id}")).await
    }
}

pub struct ItemsApi {
    client: DefarmClient,
}

impl ItemsApi {
    pub(crate) fn new(client: DefarmClient) -> Self {
        Self { client }
    }

    pub async fn list(&self, circuit_id: Option<&str>) -> Result<Vec<Item>> {
        let path = match circuit_id {
            Some(cid) => format!("/api/items?circuit_id={}", urlencoded(cid)),
            None => "/api/items".to_string(),
        };
        let resp: CollectionResponse<Item> = self.client.get(&path).await?;
        Ok(resp.into_inner())
    }

    pub async fn show(&self, id: &str) -> Result<Item> {
        self.client.get(&format!("/api/items/{id}")).await
    }

    pub async fn create_via_ingestion(
        &self,
        source_circuit_id: &str,
        items: Vec<ItemIngestionInput>,
    ) -> Result<PartnerIntakeResponse> {
        let body = json!({
            "source_circuit_id": source_circuit_id,
            "items": items,
        });
        self.client.post("/v1/partner/ingestions", &body).await
    }
}

pub struct EventsApi {
    client: DefarmClient,
}

impl EventsApi {
    pub(crate) fn new(client: DefarmClient) -> Self {
        Self { client }
    }

    pub async fn list(&self, circuit_id: Option<&str>) -> Result<Vec<Event>> {
        let path = match circuit_id {
            Some(cid) => format!("/api/events?circuit_id={}", urlencoded(cid)),
            None => "/api/events".to_string(),
        };
        let resp: CollectionResponse<Event> = self.client.get(&path).await?;
        Ok(resp.into_inner())
    }

    pub async fn show(&self, id: &str) -> Result<Event> {
        self.client.get(&format!("/api/events/{id}")).await
    }

    pub async fn add(&self, input: &EventInput) -> Result<Value> {
        self.client.post("/api/events", input).await
    }
}

pub struct DisclosuresApi {
    client: DefarmClient,
}

impl DisclosuresApi {
    pub(crate) fn new(client: DefarmClient) -> Self {
        Self { client }
    }

    pub async fn create(&self, request: &DisclosureRequest) -> Result<DisclosureResponse> {
        self.client.post("/api/disclosures", request).await
    }
}

pub struct ReceiptsApi {
    client: DefarmClient,
}

impl ReceiptsApi {
    pub(crate) fn new(client: DefarmClient) -> Self {
        Self { client }
    }

    pub async fn list(&self, circuit_id: Option<&str>) -> Result<Vec<ReceiptSummary>> {
        let path = match circuit_id {
            Some(cid) => format!("/api/receipts?circuit_id={}", urlencoded(cid)),
            None => "/api/receipts".to_string(),
        };
        let resp: CollectionResponse<ReceiptSummary> = self.client.get(&path).await?;
        Ok(resp.into_inner())
    }

    pub async fn show(&self, id: &str) -> Result<Value> {
        self.client.get(&format!("/api/receipts/{id}")).await
    }
}

fn urlencoded(s: &str) -> String {
    url::form_urlencoded::byte_serialize(s.as_bytes()).collect()
}
